third_party/move/move-binary-format/src/compatibility.rs

```diff
@@ -72,47 +77,53 @@ impl Compatibility {
-    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
-        let mut struct_and_pub_function_linking = true;
-        let mut struct_layout = true;
-        let mut friend_linking = true;
+    pub fn check(
+        &self,
+        old_module: &CompiledModule,
+        new_module: &CompiledModule,
+    ) -> PartialVMResult<()> {
+        let mut struct_and_pub_function_linking_errors = vec![];
+        let mut struct_layout_errors = vec![];
+        let mut friend_linking_errors = vec![];
 
         // module's name and address are unchanged
-        if old_module.address != new_module.address || old_module.name != new_module.name {
-            struct_and_pub_function_linking = false;
+        if old_module.address() != new_module.address() || old_module.name() != new_module.name() {
+            struct_and_pub_function_linking_errors
+                .push("module name or address changed".to_owned());
         }
 
+        let old_view = ModuleView::new(old_module);
+        let new_view = ModuleView::new(new_module);
+
         // old module's structs are a subset of the new module's structs
-        for (name, old_struct) in &old_module.structs {
-            let new_struct = match new_module.structs.get(name) {
+        for old_struct in old_view.structs() {
+            let new_struct = match new_view.struct_definition(old_struct.name()) {
                 Some(new_struct) => new_struct,
                 None => {
                     // Struct not present in new . Existing modules that depend on this struct will fail to link with the new version of the module.
                     // Also, struct layout cannot be guaranteed transitively, because after
                     // removing the struct, it could be re-added later with a different layout.
-                    struct_and_pub_function_linking = false;
-                    struct_layout = false;
+                    struct_and_pub_function_linking_errors
+                        .push(format!("struct `{}` removed", old_struct.name()));
+                    struct_layout_errors.push(format!("struct `{}` removed", old_struct.name()));
                     break;
                 },
             };
 
-            if !struct_abilities_compatibile(old_struct.abilities, new_struct.abilities)
-                || !struct_type_parameters_compatibile(
-                    &old_struct.type_parameters,
-                    &new_struct.type_parameters,
+            if !struct_abilities_compatible(old_struct.abilities(), new_struct.abilities())
+                || !struct_type_parameters_compatible(
+                    old_struct.type_parameters(),
+                    new_struct.type_parameters(),
                 )
             {
-                struct_and_pub_function_linking = false;
+                struct_and_pub_function_linking_errors.push(format!(
+                    "struct `{}` abilities or type parameters changed",
```

Comment:
Why a joint error (would be better to know which one (abilities or type params) have changed.

Reply:
It is now in the error message what has changed, as George pointed out. (Much better then before.) This simplifies the code and is also faster because errors not requested are never generated in the first place.